Set singular names for core types to pass to discovery

ardaguclu opened the pull request:
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Core types don't set their singular names for discovery endpoint(as opposed to CRDs).
This results in nondeterministic behavior when there is a CRD whose shortname is
one of core types singular name format. Thereby, when user requests resource via this short name,
it becomes unpredictable(because it is determined by the order of which one is requested first).

This PR adds singular names for all core types. As a result, core types will
have always higher precedence when their singular name format is requested
regardless of request order.

#### Which issue(s) this PR fixes:
Fixes #113227

#### Special notes for your reviewer:
In terms of backwards compatibility, this change has negative impact only on the use cases where people currently are expecting CRDs via singular form of core types. Isn't this already extreme case and incorrect expectation?
#### Does this PR introduce a user-facing change?
```release-note
None
```

ardaguclu committed:
Add integration tests for shortnames expanding to correct resources

This adds new integration tests to test shortnames and
singular names are expanding to correct resources.

In this case, core types have always higher precendence than
CRDs.

ardaguclu committed:
Introduce singularNameProvider for core types

This introduces `singularNameProvider`. This provider will be used
by core types to have their singular names are defined in discovery
endpoint. Thanks to that, core resources singular name always have
higher precedence than CRDs shortcuts or singular names.

ardaguclu committed:
run discovery tests if resources are supported

ardaguclu commented on test/cmd/discovery.sh:
```diff
@@ -55,7 +55,152 @@ run_assert_short_name_tests() {
 
   ## test if a short name is exported during discovery
-  kube::test::if_has_string "${output_message}" '{"name":"configmaps","singularName":"","namespaced":true,"kind":"ConfigMap","verbs":\["create","delete","deletecollection","get","list","patch","update","watch"\],"shortNames":\["cm"\],"storageVersionHash":'
+  kube::test::if_has_string "${output_message}" '{"name":"configmaps","singularName":"configmap","namespaced":true,"kind":"ConfigMap","verbs":\["create","delete","deletecollection","get","list","patch","update","watch"\],"shortNames":\["cm"\],"storageVersionHash":'
```
This test should fail on master without this patch.

ardaguclu commented on test/cmd/discovery.sh:
```diff
@@ -55,7 +55,152 @@ run_assert_short_name_tests() {
+  kube::test::if_has_string "${output_message}" "test-crd-example"
+
+  # test that get pod returns v1/pod instead crd
+  output_message=$(kubectl get pod)
```
This test should fail on master without this patch.

ardaguclu commented:
current failing tests don't seem to be related to the changes;
/retest

just to be sure rerun integration tests;
/test pull-kubernetes-integration

ardaguclu commented:
/retest

ardaguclu committed:
Add singular name for the rest of types

ardaguclu commented:
@deads2k @liggitt could you please take a look at this?. Thanks.

liggitt commented:
~all storage implementations already set a DefaultQualifiedResource field to the plural version of their resource... I think a peer field SingularQualifiedResource would be a clearer place to set this, making the SingularResource getter be implemented by the `Store` type

liggitt commented on pkg/registry/admissionregistration/mutatingwebhookconfiguration/storage/storage.go:
```diff
@@ -63,3 +63,10 @@ var _ rest.CategoriesProvider = &REST{}
+
+// SingularName implements the SingularNameProvider interfaces. This returns singular name of core resource.
+func (r *REST) SingularName() string {
+	return "mutatingwebhookconfiguration"
```
I think setting a peer field to `DefaultQualifiedResource` above would be clearer (e.g. `SingularQualifiedResource: admissionregistration.Resource("mutatingwebhookconfiguration"),`

(same comment applies to all)

ardaguclu replied:
Indeed that ^ would be clearer, I'll change the patch this way.

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/installer.go:
```diff
@@ -994,6 +994,9 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
 	if categoriesProvider, ok := storage.(rest.CategoriesProvider); ok {
 		apiResource.Categories = categoriesProvider.Categories()
 	}
+	if singularNameProvider, ok := storage.(rest.SingularNameProvider); ok {
```
this was the bit that we were saying should be non-optional... storage *must* implement this

liggitt replied:
having an interface is fine, just error if the storage doesn't implement the interface and we're going to surface any endpoints for it (`len(actions) > 0`)

ardaguclu replied:
it makes sense now thanks.

ardaguclu commented:
@liggitt thanks a lot for reviewing this PR.

I have one good and one bad news;
- good news is I had to touch lots of places I'm willing to learn for a long time :)
- bad news is I had to touch lots of places because subresource REST interfaces are not implicitly composed with `store`. That's why, I had to call them manually(`r.store.GetSingularName`) and that made PR bigger.

ardaguclu commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -103,6 +103,12 @@ func (r *EvictionREST) Destroy() {
 
+var _ rest.SingularNameProvider = &EvictionREST{}
+
+func (r *EvictionREST) GetSingularName() string {
```
This one is the most exceptional one. It is subresource for pod but kind is set to `Eviction`. Should the singular name be `pod` or `eviction`?

ardaguclu committed:
Use correct singular name format for subresources

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -728,6 +728,7 @@ func (r *crdHandler) getOrCreateServingInfoFor(uid types.UID, name string) (*crd
 		parameterCodec := runtime.NewParameterCodec(parameterScheme)
 
 		resource := schema.GroupVersionResource{Group: crd.Spec.Group, Version: v.Name, Resource: crd.Status.AcceptedNames.Plural}
+		singularResource := schema.GroupVersionResource{Group: crd.Spec.Group, Version: v.Name, Resource: crd.Status.AcceptedNames.Singular}
```
we have to be 100% positive _all_ existing CRDs ever created have a non-empty singular name, or come up with a fallback here, since customresource.NewStorage will panic if CompleteWithOptions returns an error

liggitt commented on pkg/registry/rbac/clusterrolebinding/policybased/storage.go:
```diff
@@ -127,3 +127,9 @@ func (s *Storage) Update(ctx context.Context, name string, obj rest.UpdatedObjec
+var _ rest.SingularNameProvider = &Storage{}
+
+func (s *Storage) GetSingularName() string {
+	return "clusterrolebinding"
```
delegate to `r.StandardStorage.GetSingularName()`

liggitt commented on pkg/registry/rbac/role/policybased/storage.go:
```diff
@@ -99,3 +99,9 @@ func (s *Storage) Update(ctx context.Context, name string, obj rest.UpdatedObjec
+var _ rest.SingularNameProvider = &Storage{}
+
+func (s *Storage) GetSingularName() string {
+	return "role"
```
delegate to `r.StandardStorage.GetSingularName()`

liggitt commented on pkg/registry/rbac/rolebinding/policybased/storage.go:
```diff
@@ -142,3 +142,9 @@ func (s *Storage) Update(ctx context.Context, name string, obj rest.UpdatedObjec
+var _ rest.SingularNameProvider = &Storage{}
+
+func (s *Storage) GetSingularName() string {
+	return "rolebinding"
```
delegate to `r.StandardStorage.GetSingularName()`

ardaguclu committed:
Remove GetSingularName for subresources

ardaguclu committed:
Use casted SingularName for rbac types

ardaguclu committed:
Implement GetSingularName for LegacyBindingREST

LegacyBindingREST is only used for `bindings` resource. It is not
a subresource and that's why it is required to implement
`GetSingularName` function. However, there is no need to implement
this function for BindingREST because it is only used for binding
subresource.

That's why, this function statically adds GetSingularName for
LegacyBindingREST.

ardaguclu committed:
minor integration test fixes and more singular resource

liggitt commented on pkg/registry/core/node/storage/storage.go:
```diff
@@ -167,6 +168,9 @@ func (r *REST) ResourceLocation(ctx context.Context, id string) (*url.URL, http.
 	return node.ResourceLocation(r, r.connection, r.proxyTransport, ctx, id)
 }
 
+// Implement ShortNamesProvider
+var _ rest.ShortNamesProvider = &REST{}
+
```
probably don't need the type assertion here if we're relying on the generic impl

liggitt commented on pkg/registry/core/pod/storage/storage.go:
```diff
@@ -288,6 +289,10 @@ func (r *LegacyBindingREST) Create(ctx context.Context, obj runtime.Object, crea
 }
 
+func (r *LegacyBindingREST) GetSingularName() string {
+	return "bindings"
```
```suggestion
	return "binding"
```

liggitt reviewed (commented):
I think we also want an integration test that enables all built-in resources and ensures we set singular names for them which match lower(kind)

Something like this:

```diff
diff --git a/test/integration/apiserver/discovery/discovery_test.go b/test/integration/apiserver/discovery/discovery_test.go
index c6b7caaa477..1906f9aa552 100644
--- a/test/integration/apiserver/discovery/discovery_test.go
+++ b/test/integration/apiserver/discovery/discovery_test.go
@@ -27,6 +27,7 @@ import (
 
 	"github.com/google/go-cmp/cmp"
 	"github.com/stretchr/testify/require"
+
 	apidiscoveryv2beta1 "k8s.io/api/apidiscovery/v2beta1"
 	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
 	apiextensions "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
@@ -690,6 +691,34 @@ func TestGroupPriorty(t *testing.T) {
 	})
 }
 
+func TestSingularNames(t *testing.T) {
+	server := kubeapiservertesting.StartTestServerOrDie(t, nil, []string{"--runtime-config=api/all=true"}, framework.SharedEtcd())
+	t.Cleanup(server.TearDownFn)
+
+	kubeClientSet, err := kubernetes.NewForConfig(server.ClientConfig)
+	require.NoError(t, err)
+
+	_, resources, err := kubeClientSet.Discovery().ServerGroupsAndResources()
+	require.NoError(t, err)
+
+	for _, rr := range resources {
+		for _, r := range rr.APIResources {
+			if strings.Contains(r.Name, "/") {
+				continue
+			}
+			if r.SingularName == "" {
+				t.Errorf("missing singularName for resource %q in %q", r.Name, rr.GroupVersion)
+				continue
+			}
+			if r.SingularName != strings.ToLower(r.Kind) {
+				t.Errorf("expected singularName for resource %q in %q to be %q, got %q", r.Name, rr.GroupVersion, strings.ToLower(r.Kind), r.SingularName)
+				continue
+			}
+		}
+	}
+
+}
+
 func makeCRDSpec(group string, kind string, namespaced bool, versions []string, categories ...string) apiextensionsv1.CustomResourceDefinitionSpec {
 	scope := apiextensionsv1.NamespaceScoped
 	if !namespaced {
```

liggitt commented:
> I think we also want an integration test that enables all built-in resources and ensures we set singular names for them which match lower(kind)

@deads2k does that restriction for built-in types match your expectation?

ardaguclu commented:
Reminder to myself: assure that aggregated discovery populates singular names.

ardaguclu committed:
Add integration test to test singularnames for all resources

ardaguclu commented:
Failure seems unrelated to this change;

/test pull-kubernetes-e2e-kind-ipv6

ardaguclu commented:
@deads2k would you mind taking a look at this one?. Thanks.

sftim commented:
> **Does this PR introduce a user-facing change?**

I think it does, even though you can only observe it via the API and not via tools. The user visible change is that the nondeterministic behavior:
> This results in nondeterministic behavior when there is a CRD whose shortname is
> one of core types singular name format. Thereby, when user requests resource via this short name,
> it becomes unpredictable(because it is determined by the order of which one is requested first).

becomes deterministic.

deads2k commented on pkg/registry/core/namespace/storage/storage.go:
```diff
@@ -94,6 +95,12 @@ func (r *REST) NamespaceScoped() bool {
 
+var _ rest.SingularNameProvider = &REST{}
+
+func (r *REST) GetSingularName() string {
```
this stands out.  Why does this one need it?

liggitt replied:
because it doesn't inline r.store and inherit the default implementation

ardaguclu replied:
it does not inherit default store implementation as said ^ https://github.com/kubernetes/kubernetes/blob/43a889fc65bc4919794ce9d43282fd2166856ad8/pkg/registry/core/namespace/storage/storage.go#L44

deads2k commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -1320,6 +1325,12 @@ func (e *Store) CompleteWithOptions(options *generic.StoreOptions) error {
 	if e.DefaultQualifiedResource.Empty() {
 		return fmt.Errorf("store %#v must have a non-empty qualified resource", e)
 	}
+	if e.SingularQualifiedResource.Empty() {
```
great.

deads2k commented:
namespace storage stood out. lgtm once I know why that needed a special case.

deads2k commented:
/lgtm
/approve

nabokihms commented on pkg/registry/authentication/selfsubjectreview/rest.go:
```diff
@@ -93,3 +93,9 @@ func (r *REST) Create(ctx context.Context, obj runtime.Object, createValidation
+var _ rest.SingularNameProvider = &REST{}
+
+func (r *REST) GetSingularName() string {
+	return "selfsubjectrulesreview"
```
@ardaguclu Hello, it seems like a bug, but I don't know why the tests are passing. I will fix this on promoting selfsubjectreviews to beta, jfyi.

nabokihms replied:
The name is `selfsubjectreview`.
